File: platform/polyglot/gemini_openai_sdk/py/gemini_openai_sdk/logger.py
"""
Logger Module - Defensive Programming Style Logging

Provides a standardized logging factory with verbose output for tracing
execution flow across the SDK.

Usage:
    from .logger import create
    logger = create("gemini_openai_sdk", __file__)

    logger.debug("Processing message", message_id=123)
    logger.info("Request completed", status=200)
    logger.warn("Rate limit approaching", remaining=10)
    logger.error("API request failed", error=str(e))
"""

import logging
import os
import sys
from pathlib import Path
from typing import Any, Optional


class SDKLogger:
    """
    SDK Logger with defensive programming patterns.

    Features:
    - Prefix output with [package:filename] context
    - Support log levels via LOG_LEVEL env var
    - Standard Console/Print interface
    - Extensible for future enhancements
    """

    LEVELS = {
        "DEBUG": logging.DEBUG,
        "INFO": logging.INFO,
        "WARN": logging.WARNING,
        "WARNING": logging.WARNING,
        "ERROR": logging.ERROR,
    }

    def __init__(
        self,
        package_name: str,
        filename: str,
        logger_instance: Optional[logging.Logger] = None,
    ):
        """
        Initialize SDK logger.

        Args:
            package_name: Name of the package (e.g., "gemini_openai_sdk")
            filename: Source file path (__file__)
            logger_instance: Optional custom logger to use instead of default
        """
        self.package_name = package_name
        self.filename = Path(filename).stem if filename else "unknown"
        self.prefix = f"[{package_name}:{self.filename}]"

        if logger_instance:
            self._logger = logger_instance
        else:
            self._logger = self._create_default_logger()

    def _create_default_logger(self) -> logging.Logger:
        """Create default logger with console handler."""
        logger = logging.getLogger(f"{self.package_name}.{self.filename}")

        # Prevent duplicate handlers
        if not logger.handlers:
            handler = logging.StreamHandler(sys.stderr)
            formatter = logging.Formatter(
                "%(asctime)s %(levelname)s %(message)s",
                datefmt="%Y-%m-%d %H:%M:%S"
            )
            handler.setFormatter(formatter)
            logger.addHandler(handler)

        # Set level from environment
        level_name = os.environ.get("LOG_LEVEL", "INFO").upper()
        level = self.LEVELS.get(level_name, logging.INFO)
        logger.setLevel(level)

        return logger

    def _format_message(self, msg: str, *args: Any, **kwargs: Any) -> str:
        """Format message with prefix and arguments."""
        formatted = f"{self.prefix} {msg}"

        # Handle positional args (printf-style)
        if args:
            try:
                formatted = formatted % args
            except (TypeError, ValueError):
                formatted = f"{formatted} {args}"

        # Handle keyword args
        if kwargs:
            kv_pairs = " ".join(f"{k}={v}" for k, v in kwargs.items())
            formatted = f"{formatted} {kv_pairs}"

        return formatted

    def debug(self, msg: str, *args: Any, **kwargs: Any) -> None:
        """Log debug message."""
        self._logger.debug(self._format_message(msg, *args, **kwargs))

    def info(self, msg: str, *args: Any, **kwargs: Any) -> None:
        """Log info message."""
        self._logger.info(self._format_message(msg, *args, **kwargs))

    def warn(self, msg: str, *args: Any, **kwargs: Any) -> None:
        """Log warning message."""
        self._logger.warning(self._format_message(msg, *args, **kwargs))

    def warning(self, msg: str, *args: Any, **kwargs: Any) -> None:
        """Log warning message (alias for warn)."""
        self.warn(msg, *args, **kwargs)

    def error(self, msg: str, *args: Any, **kwargs: Any) -> None:
        """Log error message."""
        self._logger.error(self._format_message(msg, *args, **kwargs))

    def set_level(self, level: str) -> None:
        """Set log level dynamically."""
        level_value = self.LEVELS.get(level.upper(), logging.INFO)
        self._logger.setLevel(level_value)


def create(
    package_name: str,
    filename: str,
    logger_instance: Optional[logging.Logger] = None,
) -> SDKLogger:
    """
    Create a logger instance for a module.

    Args:
        package_name: Name of the package (e.g., "gemini_openai_sdk")
        filename: Source file path (use __file__)
        logger_instance: Optional custom logger to override default

    Returns:
        SDKLogger instance with defensive logging patterns

    Example:
        from .logger import create
        logger = create("gemini_openai_sdk", __file__)

        logger.debug("enter function", param1="value")
        logger.info("operation complete", result_count=5)
        logger.error("failed", error=str(e))
    """
    return SDKLogger(package_name, filename, logger_instance)

File: platform/polyglot/gemini_openai_sdk/mjs/gemini-openai-sdk/logger.mjs
/**
 * Logger Module - Defensive Programming Style Logging
 *
 * Provides a standardized logging factory with verbose output for tracing
 * execution flow across the SDK.
 *
 * Usage:
 *   import { create } from './logger.mjs';
 *   const logger = create('gemini-openai-sdk', import.meta.url);
 *
 *   logger.debug('Processing message', { messageId: 123 });
 *   logger.info('Request completed', { status: 200 });
 *   logger.warn('Rate limit approaching', { remaining: 10 });
 *   logger.error('API request failed', { error: err.message });
 */

import { fileURLToPath } from 'url';
import { basename } from 'path';

/**
 * Log levels enum
 */
const LEVELS = {
  DEBUG: 0,
  INFO: 1,
  WARN: 2,
  WARNING: 2,
  ERROR: 3,
};

/**
 * Get current log level from environment
 */
function getLogLevel() {
  const levelName = (process.env.LOG_LEVEL || 'INFO').toUpperCase();
  return LEVELS[levelName] ?? LEVELS.INFO;
}

/**
 * Format timestamp for log output
 */
function formatTimestamp() {
  return new Date().toISOString().replace('T', ' ').split('.')[0];
}

/**
 * Format arguments for log output
 */
function formatArgs(args) {
  if (!args || (Array.isArray(args) && args.length === 0)) {
    return '';
  }

  if (typeof args === 'object' && !Array.isArray(args)) {
    return ' ' + Object.entries(args)
      .map(([k, v]) => `${k}=${JSON.stringify(v)}`)
      .join(' ');
  }

  if (Array.isArray(args)) {
    return ' ' + args.map(a => JSON.stringify(a)).join(' ');
  }

  return ' ' + String(args);
}

/**
 * SDK Logger class with defensive programming patterns.
 */
class SDKLogger {
  /**
   * @param {string} packageName - Package name for prefix
   * @param {string} filename - Source file (import.meta.url)
   * @param {object} [customLogger] - Optional custom logger instance
   */
  constructor(packageName, filename, customLogger = null) {
    this._packageName = packageName;
    this.packageName = packageName;

    // Extract filename from URL or path
    if (filename) {
      try {
        const filePath = filename.startsWith('file://')
          ? fileURLToPath(filename)
          : filename;
        this._filename = filename;  // Keep original
        this.filename = basename(filePath, '.mjs');
      } catch {
        this._filename = filename || '';
        this.filename = 'unknown';
      }
    } else {
      this._filename = '';
      this.filename = 'unknown';
    }

    this.prefix = `[${packageName}:${this.filename}]`;
    this.customLogger = customLogger;
  }

  /**
   * Format message with prefix
   */
  _formatMessage(msg, args) {
    return `${formatTimestamp()} ${this.prefix} ${msg}${formatArgs(args)}`;
  }

  /**
   * Log at specified level
   */
  _log(level, levelName, msg, args) {
    if (level < getLogLevel()) {
      return;
    }

    const formatted = this._formatMessage(msg, args);

    if (this.customLogger) {
      const method = levelName.toLowerCase();
      if (typeof this.customLogger[method] === 'function') {
        this.customLogger[method](formatted);
        return;
      }
    }

    switch (levelName) {
      case 'DEBUG':
        console.debug(`DEBUG ${formatted}`);
        break;
      case 'INFO':
        console.info(`INFO ${formatted}`);
        break;
      case 'WARN':
        console.warn(`WARN ${formatted}`);
        break;
      case 'ERROR':
        console.error(`ERROR ${formatted}`);
        break;
      default:
        console.log(`${levelName} ${formatted}`);
    }
  }

  /**
   * Log debug message
   * @param {string} msg - Message to log
   * @param {object|array} [args] - Additional arguments
   */
  debug(msg, args) {
    this._log(LEVELS.DEBUG, 'DEBUG', msg, args);
  }

  /**
   * Log info message
   * @param {string} msg - Message to log
   * @param {object|array} [args] - Additional arguments
   */
  info(msg, args) {
    this._log(LEVELS.INFO, 'INFO', msg, args);
  }

  /**
   * Log warning message
   * @param {string} msg - Message to log
   * @param {object|array} [args] - Additional arguments
   */
  warn(msg, args) {
    this._log(LEVELS.WARN, 'WARN', msg, args);
  }

  /**
   * Log warning message (alias)
   */
  warning(msg, args) {
    this.warn(msg, args);
  }

  /**
   * Log error message
   * @param {string} msg - Message to log
   * @param {object|array} [args] - Additional arguments
   */
  error(msg, args) {
    this._log(LEVELS.ERROR, 'ERROR', msg, args);
  }

  /**
   * Set log level dynamically
   * @param {string} level - Level name (DEBUG, INFO, WARN, ERROR)
   */
  setLevel(level) {
    process.env.LOG_LEVEL = level.toUpperCase();
  }
}

/**
 * Create a logger instance for a module.
 *
 * @param {string} packageName - Name of the package
 * @param {string} filename - Source file (use import.meta.url)
 * @param {object} [loggerInstance] - Optional custom logger
 * @returns {SDKLogger} Logger instance
 *
 * @example
 * import { create } from './logger.mjs';
 * const logger = create('gemini-openai-sdk', import.meta.url);
 *
 * logger.debug('enter function', { param1: 'value' });
 * logger.info('operation complete', { resultCount: 5 });
 * logger.error('failed', { error: err.message });
 */
export function create(packageName, filename, loggerInstance = null) {
  return new SDKLogger(packageName, filename, loggerInstance);
}

export { SDKLogger, LEVELS };
export default { create, SDKLogger, LEVELS };
